### PR TITLE
Expose the diff from ComparisonResult

### PR DESCRIPTION
**Pull request by Piinks**
This exposes the diff percentage of a ComparisonResult, which will enable sub-classed LocalFileComparators to set a tolerance level for golden file testing.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by CaseyHillers — APPROVED** [review]

File: packages/flutter_test/lib/src/_goldens_io.dart [hunk 1]
```diff
@@ -240,10 +245,12 @@ Future<ComparisonResult> compareLists(List<int>? test, List<int>? master) async
   }
 
   if (pixelDiffCount > 0) {
+    final double diffPercent = pixelDiffCount/totalPixels) * 100;
```
Comment (Piinks) [inline packages/flutter_test/lib/src/_goldens_io.dart:248]:
```suggestion
    final double diffPercent = pixelDiffCount / totalPixels;
```

File: packages/flutter_test/lib/src/_goldens_io.dart [hunk 2]
```diff
@@ -240,10 +245,12 @@ Future<ComparisonResult> compareLists(List<int>? test, List<int>? master) async
...
     return ComparisonResult(
       passed: false,
+      diffPercent: diffPercent,
       error: 'Pixel test failed, '
-        '${((pixelDiffCount/totalPixels) * 100).toStringAsFixed(2)}% '
+        '${diffPercent.toStringAsFixed(2)}% '
```
Comment (Piinks) [inline packages/flutter_test/lib/src/_goldens_io.dart:253]:
```suggestion
        '${(diffPercent * 100).toStringAsFixed(2)}% '
```

File: packages/flutter_test/lib/src/_goldens_io.dart [hunk 3]
```diff
@@ -170,11 +170,15 @@ class LocalComparisonOutput {
 /// [test] and [master] image bytes provided.
 Future<ComparisonResult> compareLists(List<int>? test, List<int>? master) async {
   if (identical(test, master))
-    return ComparisonResult(passed: true);
+    return ComparisonResult(
+      passed: true
```
Comment (Piinks) [inline packages/flutter_test/lib/src/_goldens_io.dart:174]:
```suggestion
      passed: true,
```

**Comment by burhanrashid52** [comment]
Can anyone help me to understand how do we set diff tolerance levels for the golden test ? I am not sure how I will use these api in my tests.
Currently I’ve golden test matcher like this.
```dart
await expectLater(find.byType(MyTestWidget),
    matchesGoldenFile('my_test_widget.png'));
 ```
How do I set a diff tolerance level for this test ?
@Piinks

**Comment by felix-barz-brickmakers** [comment]
anyone?

**Comment by fzyzcjy** [comment]
Hi, may I know what is `goldenFileKey`?

**Comment by fzyzcjy** [comment]
Seems that the code above does not work for macos. The following works:

```
MyLocalFileComparator(
      Uri.file(path.join(path.fromUri((goldenFileComparator as LocalFileComparator).basedir), 'just_arbitrary_string.dart')));
```

File: packages/flutter_test/lib/src/_goldens_io.dart [hunk 4]
```diff
@@ -240,10 +245,12 @@ Future<ComparisonResult> compareLists(List<int>? test, List<int>? master) async
   }
 
   if (pixelDiffCount > 0) {
+    final double diffPercent = pixelDiffCount / totalPixels;
     return ComparisonResult(
       passed: false,
+      diffPercent: diffPercent,
       error: 'Pixel test failed, '
-        '${((pixelDiffCount/totalPixels) * 100).toStringAsFixed(2)}% '
+        '${(diffPercent * 100).toStringAsFixed(2)}% '
```
[the remark this answers is not in the document]
Comment (matiasleyba) [reply packages/flutter_test/lib/src/_goldens_io.dart:253]: Hi @Piinks 👋🏻
Like @tp , I found it unexpected that we’re multiplying a value that’s already a percentage. Could you clarify this when you have a moment? Thanks in advance!

**Pull request merged**